alibi_detect/utils/tests/test_saving.py

```diff
@@ -1,242 +1,888 @@
+        assert cd_load.n_features == enc_dim  # encoder dim
+    assert cd_load.p_val == P_VAL
+    assert isinstance(cd_load.preprocess_fn, Callable)
+    assert cd_load.preprocess_fn.func.__name__ == 'preprocess_drift'
+    np.testing.assert_array_equal(cd.predict(X_h0)['data']['p_val'],  # only do for deterministic detectors
+                                  cd_load.predict(X_h0)['data']['p_val'])
+
+
+@pytest.mark.skipif(version.parse(scipy.__version__) < version.parse('1.7.0'),
+                    reason="Requires scipy version >= 1.7.0")
+@parametrize_with_cases("data", cases=ContinuousData, prefix='data_')
+def test_save_cvmdrift(data, preprocess_custom, tmp_path):
+    """
+    Test CVMDrift on continuous datasets, with UAE as preprocess_fn.
+
+    Detector is saved and then loaded, with assertions checking that the reinstantiated detector is equivalent.
+    """
+    # Detector save/load
+    X_ref, X_h0 = data
+    cd = CVMDrift(X_ref,
+                  p_val=P_VAL,
+                  preprocess_fn=preprocess_custom,
+                  preprocess_at_init=True,
+                  )
+    save_detector(cd, tmp_path)
+    cd_load = load_detector(tmp_path)
+
+    # Assert
+    np.testing.assert_array_equal(preprocess_custom(X_ref), cd_load.x_ref)
+    assert cd_load.x_ref_preprocessed
+    assert cd_load.n_features == LATENT_DIM
+    assert cd_load.p_val == P_VAL
+    assert isinstance(cd_load.preprocess_fn, Callable)
+    assert cd_load.preprocess_fn.func.__name__ == 'preprocess_drift'
+    np.testing.assert_array_equal(cd.predict(X_h0)['data']['p_val'],  # only do for deterministic detectors
+                                  cd_load.predict(X_h0)['data']['p_val'])
+
+
+@parametrize_with_cases("data", cases=ContinuousData, prefix='data_')
+def test_save_mmddrift(data, preprocess_custom, backend, tmp_path):
+    """
+    Test MMDDrift on continuous datasets, with UAE as preprocess_fn.
+
+    Detector is saved and then loaded, with assertions checking that the reinstantiated detector is equivalent.
+    """
+    # Detector save/load
+    X_ref, X_h0 = data
+    cd = MMDDrift(X_ref,
+                  p_val=P_VAL,
+                  backend=backend,
+                  preprocess_fn=preprocess_custom,
+                  n_permutations=N_PERMUTATIONS,
+                  preprocess_at_init=True,
+                  )
+    save_detector(cd, tmp_path)
+    cd_load = load_detector(tmp_path)
+
+    # assertions
+    np.testing.assert_array_equal(preprocess_custom(X_ref), cd_load._detector.x_ref)
+    assert cd_load._detector.x_ref_preprocessed
+    assert not cd_load._detector.infer_sigma
+    assert cd_load._detector.n_permutations == N_PERMUTATIONS
+    assert cd_load._detector.p_val == P_VAL
+    assert isinstance(cd_load._detector.preprocess_fn, Callable)
+    assert cd_load._detector.preprocess_fn.func.__name__ == 'preprocess_drift'
+#    assert cd.predict(X_h0)['data']['p_val'] == cd_load.predict(X_h0)['data']['p_val']  # Not deterministic
```

Comment:
Why does setting the seed not make this deterministic?

Reply:
Good question! I've checked, and setting all seeds (i.e. `np`, `tf` etc) doesn't guarantee determinism even on cpu. This is a known issue with the learned detectors (https://github.com/SeldonIO/alibi-detect/issues/390), but not clear why it occurs for some others. We really need to look into randomness with Tensorflow/PyTorch properly. I've made a note in the OP to open an issue about it.